usb/ehci/ehci-hub.c

```diff
@@ -0,0 +1,252 @@
+
+	if (val & PORTSC_PR)
+		status->wPortStatus |= USB_PORT_STAT_RESET;
+
+	if (val & PORTSC_PP)
+		status->wPortStatus |= USB_PORT_STAT_POWER;
+
+	if (val & PORTSC_PTC)
+		status->wPortStatus |= USB_PORT_STAT_TEST;
+
+	if (ehci->portResetChange & (1 << port))
+		status->wPortChange |= USB_PORT_STAT_C_RESET;
+
+	if ((val & PORTSC_PSPD) >> 26 == 1)
+		status->wPortStatus |= USB_PORT_STAT_LOW_SPEED;
+	else if ((val & PORTSC_PSPD) >> 26 == 2)
+		status->wPortStatus |= USB_PORT_STAT_HIGH_SPEED;
+
+	/* TODO: set indicator */
+
+	return 0;
+}
+
+
+static int ehci_setPortFeature(usb_dev_t *hub, int port, uint16_t wValue)
+{
+	hcd_t *hcd = hub->hcd;
+	volatile int *portsc = hcd->base + portsc1 + port - 1;
+	uint32_t val = *portsc;
+
+	if (port > hub->nports)
+		return -1;
+
+	val &= ~PORTSC_CBITS;
+
+	switch (wValue) {
+		case USB_PORT_FEAT_RESET:
+			ehci_resetPort(hcd, port);
+			break;
+		case USB_PORT_FEAT_SUSPEND:
+		case USB_PORT_FEAT_POWER:
+		case USB_PORT_FEAT_TEST:
+		case USB_PORT_FEAT_INDICATOR:
+			/* TODO */
+			break;
+		default:
+			return -1;
+	}
+
+	return 0;
+}
+
+
+static int ehci_clearPortFeature(usb_dev_t *hub, int port, uint16_t wValue)
+{
+	hcd_t *hcd = hub->hcd;
+	ehci_t *ehci = (ehci_t *)hcd->priv;
+	volatile int *portsc = hcd->base + portsc1 + port - 1;
+	uint32_t val = *portsc;
+
+	if (port > hub->nports)
+		return -1;
+
+	val &= ~PORTSC_CBITS;
+	switch (wValue) {
+		/* For 'change' features, ack the change */
+		case USB_PORT_FEAT_C_CONNECTION:
+			*portsc = val | PORTSC_CSC;
+			break;
+		case USB_PORT_FEAT_C_ENABLE:
+			*portsc = val | PORTSC_PEC;
+			break;
+		case USB_PORT_FEAT_C_OVER_CURRENT:
+			*portsc = val | PORTSC_OCC;
+			break;
+		case USB_PORT_FEAT_C_RESET:
+			ehci->portResetChange &= ~(1 << port);
+			break;
+		case USB_PORT_FEAT_ENABLE:
+			/* Disable port */
+			*portsc = val & ~PORTSC_ENA;
+			break;
+		case USB_PORT_FEAT_POWER:
+		case USB_PORT_FEAT_INDICATOR:
+		case USB_PORT_FEAT_SUSPEND:
+		case USB_PORT_FEAT_C_SUSPEND:
+			/* TODO */
+			break;
+		default:
+			return -1;
+	}
+
+	return 0;
+}
+
+
+static int ehci_getDesc(usb_dev_t *hub, char *buf, size_t len)
+{
+	hcd_t *hcd = hub->hcd;
+	usb_hub_desc_t *desc;
+
+	if (len < 9)
+		return -1;
+
+	desc = (usb_hub_desc_t *)buf;
+	desc->bDescLength = 9;
+	desc->bDescriptorType = USB_DESC_TYPE_HUB;
+	desc->wHubCharacteristics = 0x1;
+	desc->bPwrOn2PwrGood = 10;
+	desc->bHubContrCurrent = 10;
+	desc->bNbrPorts = *(hcd->base + hcsparams) & 0xf;
+	desc->variable[0] = 0;    /* Device not removable */
+	desc->variable[1] = 0xff; /* PortPwrCtrlMask */
+
+	return 0;
+}
+
+
+uint32_t ehci_getHubStatus(usb_dev_t *hub)
+{
+	hcd_t *hcd = hub->hcd;
+	volatile int *portsc;
+	uint32_t status = 0;
+	int i;
+
+	for (i = 0; i < hub->nports; i++) {
+		portsc = hcd->base + portsc1 + i;
+		if (*portsc & PORTSC_CSC)
+			status |= 1 << (i + 1);
+		if ((*portsc & PORTSC_CSC) && (*portsc & PORTSC_CCS) == 0)
+			phy_enableHighSpeedDisconnect(hcd, 0);
+	}
```

Comment:
Slightly faster solution, eliminates 2 extra accesses to the hardware register

```suggestion
	uint32_t status = 0;
	int i, val;

	for (i = 0; i < hub->nports; i++) {
		val = *(hcd->base + portsc1 + i);
		if (val & PORTSC_CSC)
			status |= 1 << (i + 1);
		if ((val & PORTSC_CSC) && (val & PORTSC_CCS) == 0)
			phy_enableHighSpeedDisconnect(hcd, 0);
	}
```